File: packages/guards/src/ssrf-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface SsrfDetectOptions { action: 'block' | 'warn'; }

const SSRF_PATTERNS: RegExp[] = [
  /https?:\/\/(?:127\.0\.0\.1|localhost|0\.0\.0\.0|10\.\d+\.\d+\.\d+|172\.(?:1[6-9]|2\d|3[01])\.\d+\.\d+|192\.168\.\d+\.\d+)/gi,
  /https?:\/\/\[::1?\]/gi,
  /https?:\/\/169\.254\.169\.254/gi,
  /https?:\/\/metadata\.google\.internal/gi,
  /https?:\/\/(?:\d+\.){3}\d+:\d{1,5}/gi,
  /file:\/\//gi,
  /gopher:\/\//gi,
  /dict:\/\//gi,
  /ftp:\/\/(?:127|10|192\.168|172\.(?:1[6-9]|2\d|3[01]))/gi,
];

export function ssrfDetect(options: SsrfDetectOptions): Guard {
  return {
    name: 'ssrf-detect', version: '0.1.0',
    description: 'Detect Server-Side Request Forgery (SSRF) patterns',
    category: 'security', supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];
      for (const p of SSRF_PATTERNS) {
        const re = new RegExp(p.source, p.flags);
        const m = re.exec(text);
        if (m) matched.push(m[0]);
      }
      const triggered = matched.length > 0;
      return {
        guardName: 'ssrf-detect', passed: !triggered,
        action: triggered ? options.action : 'allow',
        message: triggered ? `SSRF pattern: "${matched[0]}"` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched, reason: 'Text contains URLs targeting internal/metadata services' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/ssrf_detect.py
"""Detects Server-Side Request Forgery (SSRF) patterns."""

import re
import time
from typing import List

from open_guardrail.core import GuardResult

_PATTERNS = [
    re.compile(r"https?://(?:localhost|127\.0\.0\.1|0\.0\.0\.0|10\.\d+\.\d+\.\d+|172\.(?:1[6-9]|2\d|3[01])\.\d+\.\d+|192\.168\.\d+\.\d+)", re.I),
    re.compile(r"https?://\[?::1\]?", re.I),
    re.compile(r"https?://169\.254\.169\.254", re.I),
    re.compile(r"https?://metadata\.google\.internal", re.I),
    re.compile(r"\bfile://", re.I),
    re.compile(r"\bgopher://", re.I),
    re.compile(r"\bdict://", re.I),
    re.compile(r"https?://.*@.*internal", re.I),
]


class _SsrfDetect:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "ssrf-detect"
        self.action = action

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matched: List[str] = []
        for p in _PATTERNS:
            m = p.search(text)
            if m:
                matched.append(m.group()[:60])
        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="ssrf-detect",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message="SSRF pattern detected" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matched": matched} if triggered else None,
        )


def ssrf_detect(*, action: str = "block") -> _SsrfDetect:
    return _SsrfDetect(action=action)
